Tests/Fakes/FakeURLSession.swift

```diff
@@ -0,0 +1,44 @@
+import Foundation
+
+/// A fake URLSession that will return data tasks which will call completion handlers with the given parameters.
+public class FakeURLSession: URLSession {
+  /// Fields that will be returned in the next URLSessionDataTask.
+  public var urlResponse: URLResponse?
+  public var data: Data?
+  public var error: Error?
+
+  /// Initialize a new FakeURLSession.
+  ///
+  /// - Parameters:
+  ///   - string: A utf-8 encoded string that will be encoded as data in the next request.
+  ///   - urlResponse: The url response to serve in the next request.
+  ///   - error: The error to serve in the next request.
+  public convenience init(string: String, urlResponse: HTTPURLResponse?, error: Error?) {
```

Comment:
What's this `convenience` init?

An override if the first parameter is `string`?

Reply:
`convenience` means that this initializer will call another initializer. We call this a convenience initializer. We call the initializer who does't have `convenience` in it a 'designated initializer`. 

Convenience initializers:
- cannot call super class initializers
- must call a designated initializer

Designated Initializers:
- cannot call convenience initializer
- can call a super class initializer

-----

Why is this a convenience initializer? 

The first parameter we want is a `Data` object. `Data` is just a wrapper classer for a collection of bytes, so it can be virtually anything. 

However, since most data we get back from web requests is just utf-8 encoded text and clients likely want to to work with text rather than bytes, this convenience initializer knows how to transform a string into bytes and serve that to a designated intializer.

The effect of this is that we write:
```
    let data = string.data(using: .utf8)
```
once in our class, rather than all of the clients of this class having to write this code every time they want to fake a string response.

Reply:
Great explanation. Thanks for the Swift tutorial!

Reply:
yeah this is gold